handlers_test.go

```diff
@@ -91,3 +92,111 @@ func Test_GivenCorrectId_HandlerReturnsDeviceObject(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, expected, result)
 }
+
+func Test_GetAllDevicesHandler_GivenWrongInput_HandlerReturns400(t *testing.T) {
+	r := newRouter(NewService(&mockDao{}))
+	mockServer := httptest.NewServer(r)
+
+	urls := []string{"/devices?limit=a", "/devices?page=a",
+		"/devices?limit=1&page=a", "/devices?limit=-1",
+		"/devices?page=-1"}
+	for _, url := range urls {
```

Comment:
This approach has one little flaw: When something breaks you have to read through the test to know which case actually failed.

Golang's test framework favours table driven tests for such situations: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Reply:
I even tend to use `map[string]struct` instead of `[]struct` so that I can name every single case.